lib/limiter/connlimiter.go

```diff
@@ -20,47 +20,64 @@ package limiter
 
 import (
 	"net/http"
+	"strings"
 	"sync"
 
-	"github.com/gravitational/oxy/connlimit"
-	"github.com/gravitational/oxy/utils"
 	"github.com/gravitational/trace"
-	log "github.com/sirupsen/logrus"
 )
 
-// ConnectionsLimiter is a network connection limiter and tracker
+// ConnectionsLimiter is a network connection limiter.
 type ConnectionsLimiter struct {
-	*connlimit.ConnLimiter
 	maxConnections int64
 
+	next http.Handler
+
 	sync.Mutex
 	connections map[string]int64
 }
 
 // NewConnectionsLimiter returns new connection limiter, in case if connection
 // limits are not set, they won't be tracked
-func NewConnectionsLimiter(config Config) (*ConnectionsLimiter, error) {
-	limiter := ConnectionsLimiter{
-		maxConnections: config.MaxConnections,
+func NewConnectionsLimiter(maxConnections int64) (*ConnectionsLimiter, error) {
+	// TODO: remove error from return type
+	return &ConnectionsLimiter{
+		maxConnections: maxConnections,
 		connections:    make(map[string]int64),
+	}, nil
+}
+
+// Wrap wraps an HTTP handler.
+func (l *ConnectionsLimiter) Wrap(h http.Handler) {
+	l.next = h
```

Comment:
I don't like this approach, a second call to Wrap is going to overwrite the first one and we might do that by accident, we could return a `http.Handler` from `Wrap` instead (with an inline `http.HandlerFunc` maybe):
```go
func (l *ConnectionsLimiter) Wrap(next http.Handler) http.Handler {
  return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    token, _, err := net.SplitHostPort(r.RemoteAddr)
    ...
    if err := l.AcquireConnection(token); err != nil {
      ...
    }
    defer l.ReleaseConnection(token)
    next.ServeHTTP(w, r)
  })
}
```

Reply:
All right then my comment is invalid -  Sorry for confusion.